core/services/database/WordPressOption.php

```diff
@@ -32,36 +34,47 @@ abstract class WordPressOption
     /**
      * @var boolean
      */
-    private $autoload;
+    private $autoload = false;
 
     /**
      * @var mixed
      */
-    private $default_value;
+    private $default_value = null;
 
     /**
      * @var string
      */
-    private $option_name;
+    private $option_name = '';
 
     /**
      * @var mixed
      */
     private $value = WordPressOption::NOT_SET_YET;
 
+    /**
+     * @var OptionEngine
+     */
+    private $option_engine;
+
 
     /**
      * WordPressOption constructor.
      *
-     * @param bool   $autoload
-     * @param mixed  $default_value
      * @param string $option_name
+     * @param mixed  $default_value
+     * @param bool   $autoload				if true, will load the option on EVERY request
+     * @param bool 	 $is_network_option		if true, will save the option to the network as opposed to the current blog
      */
-    public function __construct(string $option_name, $default_value, bool $autoload = false)
-    {
+    public function __construct(
+        string $option_name,
+        $default_value,
+        bool $autoload = false,
+        bool $is_network_option = false
+    ) {
         $this->setAutoload($autoload);
         $this->setDefaultValue($default_value);
         $this->setOptionName($option_name);
+        $this->option_engine = new OptionEngine($is_network_option);
```

Comment:
Nice idea 👍

Reply:
this is following the design mantra of:

> ### isolate the thing that changes

I _could_ have just embedded all of the logic that's in the `OptionEngine` class within this class, but that would have required a conditional anytime we needed to use one of the `*_option()` functions. Duplicate conditionals within a class are a sure sign that there is code that can be extracted. 

Another way I could have done it would to have had two `OptionEngine` classes like:

- `BlogOption`
- `NetworkOption`

and then have a single conditional in WordPressOption for choosing between them:

```php
$this->option_engine = $is_network_option ? new NetworkOption() : new BlogOption();
```

but the `OptionEngine` class was simple enough that I didn't feel the need to have to split it up anymore